### PR TITLE
Fix calculation of link-only sample size for u-training

### LINKED ISSUE #1000: SQLite random SQL doesn't allow customised `unique_id_column_name`
### What happens?

If you create a SQLite linker with a non-default option of `unique_id_column_name` then `linker.estimate_u_using_random_sampling` (and probably `linker.cluster_studio_dashboard`) break with `sqlite3.OperationalError: no such column: unique_id`.

Issue is hard-coded column name [here](https://github.com/moj-analytical-services/splink/blob/master/splink/sqlite/sqlite_linker.py#L161-L164).

### To Reproduce

```python
import pandas as pd
import numpy as np

import sqlite3
from splink.sqlite.sqlite_linker import SQLiteLinker
import splink.sqlite.sqlite_comparison_library as cl

fnames = ("shore", "flerk", "blag", "lunk")
snames = ("frett", "plood", "nepsy")
n_rows = 100
df = pd.DataFrame(
    {
        "person_id": range(0, n_rows),
        "fname": np.array(list(map(lambda x: np.random.choice(fnames, 1)[0], range(0, n_rows)))),
        "sname": np.array(list(map(lambda x: np.random.choice(snames, 1)[0], range(0, n_rows)))),
    }
)

con = sqlite3.connect(":memory:")
df.to_sql("people", con, if_exists="replace")

settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [],
    "unique_id_column_name": "person_id",
    "comparisons": [
        cl.exact_match("fname"),
        cl.exact_match("sname"),
    ],
}

linker = SQLiteLinker("people", settings_dict=settings, connection=con)
# some sample of less than 100% of possible links:
linker.estimate_u_using_random_sampling(100)
```

### OS:

OSX

### Splink version:

3.5.4

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### LINKED ISSUE #1232: U-train random sampling link-only not working properly for SQLite
### What happens?

Results are not correct in a link-only job when training u using SQLite as backend.

The issue seems to be that for that backend we are using the sample size rather than proportion, and an incorrect one is passed to the function.

### To Reproduce

Run:
```sh
pytest -W ignore -m sqlite tests/test_u_train.py::test_u_train_link_only_sample
```
using changes on branch #1205.

The test is not super clear and may benefit from being rewritten, but highlights the issue.

### OS:

MacOS

### Splink version:

3.8.1

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by ADBond**
### Type of PR

- [x] BUG
- [ ] FEAT
- [ ] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
Closes #1232
Closes #1000

Also solves an undocumented issue where `cluster_studio_dashboard` would fail for sqlite if using random clusters.

These issues also apply to `postgres` - this fixes the equivalent of #1232 there, with the rest just requiring an equivalent update to `PostgresLinker._random_sample_sql`.

### Give a brief description for the solution you have provided
The calculation in `estimate_u_using_random_sample` for link-only jobs gave the wrong `sample_size` (i.e. the number of source nodes to sample). This didn't affect `spark` or `duckdb` as they used the `proportion` which _was_ calculated correctly.

I have:
* corrected this calculation
* fixed a logical error in sqlite `_random_sample_sql` which meant we sampled the wrong number when there was more than one data-set (by including the source dataset in the effective id)
* allowed sqlite `_random_sample_sql` to find the unique id column + source dataset column from settings, rather than being hard-coded (in the former case)
* added a test covering some lesser-used settings options (relevant in particular for this PR: `unique_id_column_name`)
* now pass table (being sampled from) and unique id to `_random_sample_sql`, as we need this info for `sqlite`, as this SQL is also used for building the cluster studio
* moved the base `_random_sample_sql` from `SplinkDataFrame` (where it was unused) to `Linker` (to help clue subclasses)

### PR Checklist

- [N/A] Added documentation for changes
- [N/A] Added feature to example notebooks at tutorials in [splink_demos](https://github.com/moj-analytical-services/splink_demos) (if appropriate)
- [x] Added tests (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

**Comment by ADBond** [comment]
☝️ investigating test issue - seems that there is a problem when running on pyspark version in CI - version 3.4.0 works fine locally

File: tests/test_u_train.py [hunk 1]
```diff
@@ -130,8 +131,9 @@ def test_u_train_link_only_sample(test_helpers, dialect):
     linker = helper.Linker([df_l, df_r], settings, **helper.extra_linker_args())
     linker.debug_mode = True
     linker.estimate_u_using_random_sampling(max_pairs=max_pairs)
-    linker._settings_obj.comparisons[0]
+    # linker._settings_obj.comparisons[0]
```
Comment (ThomasHepworth) [inline tests/test_u_train.py:134]: rm if you're commenting this out

**Comment by ADBond** [comment]
@ThomasHepworth I have made an identical change to the postgres linker now that #1251 is merged.

Not sure why the autoblack workflow is failing - looks like maybe an issue with cached env?

**Comment by ADBond** [comment]
(it was - have reset the lint cache and worked okay)

**Pull request merged**